Email Confirmation Changes

aaronskiba opened the pull request:
Fixes * 

Changes proposed in this PR:
- Re-implement email confirmation via Devise's `:confirmable` module
- Add rake task that after execution, all non-superadmins will have to confirm their email to use the app
- Customise UX flow for existing, unconfirmed users

aaronskiba committed:
Add `:confirmable` to included devise modules

aaronskiba committed:
Replace .yml value from DMPOnline to DMP Assistant

The corresponding value for `devise.mailer.confirmation_instructions.subject` in `config/locales/translation.fr-CA.yml` specifies "Assistant PGD", but maybe it should be verified as well.

Note: I don't think this is the proper way to update the locales files. Rather, I believe only `config/locales/en.yml` should be directly modified. However, https://github.com/portagenetwork/roadmap/issues/920 is currently preventing that.

aaronskiba committed:
TEMP: customise devise.failure.unconfirmed values

This commit customises Devise's default value for the key `devise.failure.unconfirmed`. The value adds a link to the email confirmation page.
Note: The values for this key in both the `...en-CA.yml` and `...fr-CA.yml` files were provided solely by me and need to be approved/finalised by The Alliance

aaronskiba committed:
Improve UX flow for existing unconfirmed users

This commit seeks to improve the UX flow for existing users with unconfirmed emails.

By default, when using Devise's `:confirmable` module, confirmation instructions are sent to the user at the time of account creation. This will be perfect for newly created users, but poses some challenges for existing users.

A "first category" of these existing users would be those who were created while `:confirmable` was absent from the codebase. These users are unconfirmed and a confirmation token was never generated for them.

A "second category" of these existing users are those that are unconfirmed but do have an existing confirmation token. Since `:confirmable` has been absent from the codebase for years, the most recent `confirmation_sent_at` value is quite old (`Mon, 22 Feb 2021 15:18:24.000000000 UTC +00:00`).

The code changes in `app/controllers/sessions_controller.rb` and `app/models/user.rb` improves the UX flow for users in the "first category". Rather than requiring the user to manually request new confirmation instructions, these changes auto-send the email when these users first attempt to sign in.

`lib/tasks/dmp_assistant_upgrade.rake` enables this improved UX flow for the aforementioned "second category" of users. `def handle_unconfirmed_users_with_outstanding_invitations` queries for these unconfirmed users and then updates both `confirmation_token` and `confirmation_sent_at` to null. As a result, these users now meet the criteria for the "first category" and can also make use of the improved UX flow.

aaronskiba committed:
Handle confirmation for SSO created users

An "SSO created user" is a user whose account was created via the `Sign in with institutional or social ID` button. This occurs when the chosen SSO email is neither already linked to an existing account, nor does there exist a user account with the same email as the chosen SSO supplied email.

In order to sign in via SSO, the user must provide the password/credentials for the SSO account that they have chosen. This could be considered a form of email confirmation. And because the user's account email will be set to match this SSO email, it follows that email confirmation via Devise's `:confirmable` module is redundant and can be bypassed.

Taking the above reasoning into account, `app/models/user.rb` sets `confirmed_at: Time.now` for any future SSO created accounts.

`lib/tasks/dmp_assistant_upgrade.rake` sets `lib/tasks/dmp_assistant_upgrade.rake` for any "existing SSO created accounts". Specifically, to be an "existing SSO created account", BOTH of the following conditions must be met:
1. An Identifier corresponding to the "openid_connect" `IdentifierScheme` exists for the user
 2. For the identifier and corresponding user found in 1., the values of `identifier.created_at` and `user.created_at` are within 1 second of each other.

aaronskiba committed:
Re-add SSO linking capability for signed-out users

This commit re-allows existing users to link via SSO while signed out.  This functionality  was previously removed via commit e112202fa.

The logic works as follows:
1) A user clicks the "Sign in with institutional or social ID" button
2) Within CILogon, the user chooses an email to sign in with. For this particular functionality to be used, the following conditions must both be true for the chosen email:
  a) It is not currently linked to any user accounts
  b) It matches an existing user.email in the database
3) If the matching user.email is confirmed, then the SSO email is linked to the user account and the user is signed-in. Otherwise, the user remains signed out and a flash notice is rendered indicating that the user.email requires confirmation.

aaronskiba committed:
Refactor / Make Rubocop happy

aaronskiba committed:
Adapt existing tests to `:confirmable` re-addition

`config/environments/test.rb`
- Add "SMTP From address" to mock Devise's sending of confirmation instructions email at time of account creation

`spec/factories/users.rb`
- Set `confirmed_at { Time.now }` in user factory

`spec/features/registrations_spec.rb`
- Change expected path to `root_path`. This is because the user should not be able to access `plans_path` until after they confirm their email
- Add check to verify that the confirmation-related flash message was sent

`spec/integration/openid_connect_sso_spec.rb`
- Removing `it 'does not create SSO link when user is signed out and SSO email is an existing account email'` test. Commit 965c3365b  ("Re-add SSO linking capability for signed-out users") updated the SSO behaviour in a manner that this test is no longer relevant.

aaronskiba committed:
Reset `:confirmable` cols for all non-superusers

- These changes replace the changes made to `lib/tasks/dmp_assistant_upgrade.rake` in commits 6087aeed1340de9957610214135abab8fc2502ec and 465d6712a069294b43748ee3b857204bb256652a. We are now setting `confirmed_at`, `confirmation_token`, and `confirmation_sent_at` to nil for all users. (the rake task subsequently sets `confirmed_at = Time.now` for all superusers).

aaronskiba committed:
Reuse `signed_up_but_unconfirmed` flash message

This commit changes the flash message we are using for our custom UX flow. Now, rather than a custom message, we are reusing `devise.registrations.signed_up_but_unconfirmed`.

lagoan commented on app/models/user.rb:
```diff
@@ -197,7 +197,10 @@ def self.create_from_provider_data(provider_data)
+      password: Devise.friendly_token[0, 20],
+      # provider_data.info.email comes from CILogon sign-in, which requires email confirmation
+      # It follows that we can set `confirmed_at: Time.now` and bypass Devise's email confirmation step
+      confirmed_at: Time.now
```
We should follow devise behaviour and have the confirmed_at value set as  `confirmed_at: Time.now.to_utc`

https://github.com/heartcombo/devise/blob/72884642f5700439cc96ac560ee19a44af5a2d45/lib/devise/models/confirmable.rb#L87

aaronskiba replied:
Cool, I was unsure about this one. I'm doing a little more testing, and we might be safe. Although the output of `Time.now` reflects my current timezone, Whether I try saving `Time.now` or `Time.now.utc`, it seems to always be saved to the DB in its UTC form:

```
3.1.4 :001 > user = User.last
  User Load (3.2ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" DESC LIMIT $1  [["LIMIT", 1]]
 => #<User id: 26480, firstname: "test", surname: "test", email: "helloworld@abc.com", created_at: ...

3.1.4 :002 > user.updated_at
 => Thu, 10 Oct 2024 20:21:56.229255000 UTC +00:00

3.1.4 :003 > user.update(updated_at: Time.now)
  TRANSACTION (2.5ms)  BEGIN
  Language Load (2.2ms)  SELECT "languages".* FROM "languages" WHERE "languages"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]                                                                        
  Org Load (3.1ms)  SELECT "orgs".* FROM "orgs" WHERE "orgs"."id" = $1 LIMIT $2  [["id", 59], ["LIMIT", 1]]                                                                                           
  User Update (3.7ms)  UPDATE "users" SET "updated_at" = $1 WHERE "users"."id" = $2  [["updated_at", "2024-10-10 20:23:35.082699"], ["id", 26480]]                                                    
  TRANSACTION (13.9ms)  COMMIT                                                               
 => true

3.1.4 :004 > user.reload
  User Load (2.5ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 26480], ["LIMIT", 1]]                                                                                    
 => #<User id: 26480, firstname: "test", surname: "test", email: "helloworld@abc.com", created_at: "2024-10-10 18:02:13.546883000 +0000", updated_at: "2024-10-10 20:23:35.082699000 +0000", other_organisation: nil, dmponline3: nil, accept_terms: true, org_id: 59, api_token: nil, language_id: 1, recovery_email: nil, active: true, department_id: nil, last_api_access: nil, prefs: nil>

3.1.4 :005 > user.updated_at
 => Thu, 10 Oct 2024 20:23:35.082699000 UTC +00:00 
3.1.4 :006 > user.update(updated_at: Time.now.utc)
  TRANSACTION (2.9ms)  BEGIN
  Language Load (2.7ms)  SELECT "languages".* FROM "languages" WHERE "languages"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]                                                                                    
  Org Load (2.3ms)  SELECT "orgs".* FROM "orgs" WHERE "orgs"."id" = $1 LIMIT $2  [["id", 59], ["LIMIT", 1]]                                                                                                       
  User Update (2.6ms)  UPDATE "users" SET "updated_at" = $1 WHERE "users"."id" = $2  [["updated_at", "2024-10-10 20:24:34.814186"], ["id", 26480]]                                                                
  TRANSACTION (13.4ms)  COMMIT                                                                           
 => true
 
3.1.4 :007 > user.reload
  User Load (3.2ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 26480], ["LIMIT", 1]]                                                                                                
 => #<User id: 26480, firstname: "test", surname: "test", email: "helloworld@abc.com", created_at: "2024-10-10 18:02:13.546883000 +0000", updated_at: "2024-10-10 20:24:34.814186000 +0000", other_organisation: nil, dmponline3: nil, accept_terms: true, org_id: 59, api_token: nil, language_id: 1, recovery_email: nil, active: true, department_id: nil, last_api_access: nil, prefs: nil>
 
3.1.4 :008 > user.updated_at
 => Thu, 10 Oct 2024 20:24:34.814186000 UTC +00:00
```

lagoan replied:
I am glad you checked that. We should still follow the gem's behaviour so we can catch any changes regardless.

lagoan replied:
We should follow the original gem and use utc time. ``confirmed_at: Time.now.to_utc``

aaronskiba replied:
Sounds good. Sorry, still need to make these changes soon here.

lagoan commented on config/locales/translation.en-CA.yml:
```diff
@@ -228,11 +228,13 @@ en-CA:
       unauthenticated: You need to sign in or sign up before continuing.
-      unconfirmed: You have to confirm your account before continuing.
+      unconfirmed: |
+        You have to confirm your account before continuing.
+        <a href="/users/confirmation/new" class="a-orange">(Click to request a new confirmation email)</a>
```
Right now we are sending the email verification message sent when the user tries to log in and is not verified. What do you think about letting the user know about this sent email in this message instead of asking the user to request the email confirmation again ?

aaronskiba replied:
Good question. This flash message is actually only rendered when the user already has an outstanding confirmation. An additional confirmation email will not be auto-sent in this case and they will have to manually request a new one.

If a user is unconfirmed and has no outstanding confirmation token, then a different message will render after they attempt to sign in (specifically, `devise.registrations.signed_up_but_unconfirmed`).

```
3.1.4 :001 > I18n.t('devise.registrations.signed_up_but_unconfirmed')
 => "A message with a confirmation link has been sent to your email address. Please open the link to activate your account. If you do not receive the confirmation email, please check your spam filter." 
```

```ruby
# app/models/user.rb
def confirmation_instructions_handled?
  confirmed? || confirmation_token.present?
end
```

```ruby
# app/controllers/sessions_controller.rb
unless existing_user.confirmation_instructions_handled?
  handle_confirmation_instructions(existing_user)
  return
end

def handle_confirmation_instructions(user)
  user.send_confirmation_instructions
  flash[:notice] = _('A confirmation email has been sent to you. Please check your inbox to confirm your account.')
  flash[:notice] = I18n.t('devise.registrations.signed_up_but_unconfirmed')
  redirect_to root_path
end
```

lagoan commented on config/locales/translation.en-CA.yml:
```diff
@@ -228,11 +228,13 @@ en-CA:
     mailer:
       confirmation_instructions:
-        subject: Confirm your DMPonline account
+        subject: Confirm your DMP Assistant account
```
Good catch for this change !

lagoan commented on config/locales/translation.fr-CA.yml:
```diff
@@ -242,7 +242,9 @@ fr-CA:
-      unconfirmed: Vous devez confirmer votre compte pour continuer.
+      unconfirmed: |
+        Vous devez confirmer votre compte pour continuer.
+        <a href="/users/confirmation/new" class="a-orange">(Cliquez pour demander un nouvel e-mail de confirmation)</a>
```
Same as above to let the user know about the email that was already sent.

lagoan commented on spec/factories/users.rb:
```diff
@@ -60,6 +60,7 @@
     email        { Faker::Internet.unique.email }
     password     { 'password' }
     accept_terms { true }
+    confirmed_at { Time.now }
```
Same as above to use ``confirmed_at { Time.now.to_utc }``

lagoan reviewed (changes requested):
Good work all around ! I added just some comments on minor changes.

aaronskiba committed:
Add tests for custom email confirmation UX flow

- These tests are meant to confirm that the desired behaviour is encountered via our custom email confirmation UX flow.
  - For more info, See comments for `app/controllers/concerns/email_confirmation_handler.rb`

aaronskiba committed:
Enable custom confirmation UX flow for SSO sign in

This commit takes the idea in commit 6087aeed1340de9957610214135abab8fc2502ec and applies it to users signing in via SSO. Via the addition of `app/controllers/concerns/email_confirmation_handler.rb`, some refactoring has been applied here as well.

aaronskiba committed:
Refactor: Move logic to `EmailConfirmationHandler`

aaronskiba committed:
Refactor / Make rubocop happy

Some refactoring was performed in `def openid_connect` to  avoid adding `Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity` to `rubocop:disable`.

aaronskiba committed:
Add test for SSO linking of signed-out users

aaronskiba committed:
Test custom I18n strings / override default_locale

In order to properly test the values of our customised `I18n.t` values (specifically, `I18n.t('devise.failure.unconfirmed')` and `I18n.t('devise.registrations.signed_up_but_unconfirmed')`), we need to set `I18n.default_locale` to match our app's default locale.

NOTE: These changes only set `I18n.default_locale = :'en-CA'` for the duration of the email confirmation tests. However, it may be best to apply this update globally throughout the tests.

aaronskiba committed:
Refactor EmailConfirmationHandler

- Rename `def missing_confirmation_instructions_handled?(user)` to `def confirmation_instructions_missing_and_handled?(user)` and add comments to where it is called for easy reference
- Move `def confirmed_or_has_confirmation_token?` from User model to `EmailConfirmationHandler`. This check is only accessed by the concern, and it seems easiest to make it a private method and encapsulate the logic there.

lagoan reviewed (changes requested):
Awesome work. Just a minor comment on how we are setting the time.

aaronskiba committed:
Update `Time.now` to `Time.now.utc`

aaronskiba committed:
Merge branch 'integration' into aaron/add-email-confirmation

aaronskiba committed:
Update CHANGELOG.md (and fix comment)

aaronskiba committed:
Add rake task for openid_connect / CILogon cleanup

Intended to resolve this issue: https://github.com/portagenetwork/roadmap/issues/912
The following PR was used as a reference: https://github.com/portagenetwork/roadmap/pull/915

Co-Authored-By: Omar Rodriguez Arenas <1537909+lagoan@users.noreply.github.com>

aaronskiba committed:
Make Rubocop happy

aaronskiba committed:
Merge branch 'integration' into aaron/add-email-confirmation

aaronskiba committed:
Add finalised email confirmation translations

- Updated via `bundle exec rake translation:sync` with prod key.
- These changes finalise some of the initial changes we made for testing in commit 6c41bd7d2bb44a4e632330380691fe82c0fc020d

aaronskiba committed:
Update test to match updated translation

- The corresponding translation was commited via the following commit: 09e731e4ebf1c8d426ad9981871531c8d351d26d

aaronskiba committed:
Update CHANGELOG.md

lagoan reviewed (dismissed):
![LGTM](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExcjYzcm4wcnJoNnA0dmg5ZnNoNGZjMWdhempiNXJkbG9odzVsZm5oZiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/5USTijryafZEQ/giphy.gif)

aaronskiba committed:
Merge pull request #944 from portagenetwork/aaron/issues/912

Add rake task for `openid_connect` / CILogon cleanup

lagoan reviewed (approved):
LGTM!
